rust/src/optimize.rs

```diff
@@ -0,0 +1,356 @@
+use parquet::file::serialized_reader::{SerializedFileReader, SliceableCursor};
+
+use crate::action::{self, Action};
+use crate::parquet::file::reader::FileReader;
+use crate::writer::utils::PartitionPath;
+use crate::writer::DeltaWriter;
+use crate::writer::DeltaWriterError;
+use crate::writer::RecordBatchWriter;
+use crate::{DeltaDataTypeLong, DeltaTable, DeltaTableError, PartitionFilter};
+
+#[derive(Default, Debug, PartialEq)]
+///Metrics from Optimize
+pub struct Metrics {
+    ///Number of optimized files added
+    pub num_files_added: u64,
+    ///Number of unoptimized files removed
+    pub num_files_removed: u64,
+    ///Detailed metrics for the add operation
+    pub files_added: MetricDetails,
+    ///Detailed metrics for the remove operation
+    pub files_removed: MetricDetails,
+    ///Number of partitions that had at least one file optimized
+    pub partitions_optimized: u64,
+    ///TODO: The number of batches written
+    pub num_batches: u64,
+    ///How many files were considered during optimization. Not every file considered is optimized
+    pub total_considered_files: usize,
+    ///How many files were considered for optimization but were skipped
+    pub total_files_skipped: usize,
+    ///The order of records from source files is preserved
+    pub preserve_insertion_order: bool,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+///Statistics on files for a particular operation
+/// Operation can be remove or add
+pub struct MetricDetails {
+    ///Maximum file size of a operation
+    pub min: DeltaDataTypeLong,
+    ///Minimum file size of a operation
+    pub max: DeltaDataTypeLong,
+    ///Average file size of a operation
+    pub avg: f64,
+    ///Number of files encountered during operation
+    pub total_files: usize,
+    ///Sum of file sizes of a operation
+    pub total_size: DeltaDataTypeLong,
+}
+
+impl Default for MetricDetails {
+    fn default() -> Self {
+        MetricDetails {
+            min: DeltaDataTypeLong::MAX,
+            max: 0,
+            avg: 0.0,
+            total_files: 0,
+            total_size: 0,
+        }
+    }
+}
+
+///Optimize a Delta table with given options
+///
+/// If a target file size is not provided then `delta.targetFileSize` from the
+/// table's configuration is read. Otherwise a default value is used
+#[derive(Default)]
+pub struct Optimize<'a> {
+    filters: &'a [PartitionFilter<'a, &'a str>],
+    target_size: Option<i64>,
+}
+
+impl<'a> Optimize<'a> {
+    ///Only optimize files that return true for the specified partition filter
+    pub fn filter(mut self, filters: &'a [PartitionFilter<'a, &'a str>]) -> Self {
+        self.filters = filters;
+        self
+    }
+
+    ///Set the target file size
+    pub fn target_size(mut self, target: DeltaDataTypeLong) -> Self {
+        self.target_size = Some(target);
+        self
+    }
+
+    ///Perform the optimization. On completion, a summary of how many files were added and deleted is returned
```

Comment:
nit: It's probably clearer to say "removed" rather than "deleted", to make clear the underlying files aren't deleted.
```suggestion
    /// Perform the optimization. On completion, a summary of how many files were added and removed is returned
```